bw_check_commit_hook/__init__.py

```diff
@@ -0,0 +1,205 @@
+        config['errors_ignored'].extend(config.get("ERRORS_IGNORED", []))
+
+        patch_files = [
+            p
+            for p in commit_files
+            if _is_relative_to(p, config_dir)
+            and p not in config.get("IGNORED_FILES", [])
+            and not p.is_symlink()
+        ]
+        if patch_files:
+            commit_patch = _get_patch(running_in_ci, patch_files)
+            _run_checkpatch(commit_patch, errors, **config)
+
+    if errors:
+        sys.exit(1)
+
+def _get_commit_files_from_log() -> List[Path]:
+    cmd = ["git", "diff-tree", "--no-commit-id", "--name-only", "HEAD", "-r"]
+    log_output = check_output(cmd).decode().rstrip("\n")
+    return [Path(file_path) for file_path in log_output.split("\n")]
+
+
+def _get_patch(running_in_ci: bool, patch_files: List[Path]) -> str:
+    if running_in_ci:
+        return check_output(
+            [
+                "git",
+                "format-patch",
+                "--stdout",
+                "-1",
+                "HEAD",
+                "--",
+                *patch_files,
+            ]
+        ).decode()
+    commit_msg = _read_commit_msg_from_file()
+    return _diff_to_patch(patch_files, commit_msg)
+
+
+def _diff_to_patch(patch_files: List[Path], commit_msg: str) -> str:
+    diff = check_output(
+        ["git", "diff", "--cached", "--", *patch_files]
+    ).decode()
+    return PATCH_TEMPLATE.format(
+        diff=diff,
+        commit_msg=commit_msg,
+        current_datetime=datetime.now().strftime("%a, %d %b %Y %H:%M:%S %z"),
+    )
+
+
+def _is_relative_to(a_path: Path, another_path: Path) -> bool:
+    # adapted from Python 3.9+, where this is introduced as a Path method
+    try:
+        a_path.relative_to(another_path)
+        return True
+    except ValueError:
+        return False
+
+
+def _read_commit_msg_from_file() -> str:
+    commit_msg_file = Path(".git") / "COMMIT_EDITMSG"
+    with commit_msg_file.open("r") as file:
+        commit_msg = file.read()
+    return commit_msg
+
+
+def _run_checkpatch(
+    commit_patch: str,
+    errors: DefaultDict[str, List],
+    errors_enabled: Optional[List] = None,
+    errors_ignored: Optional[List] = None,
+    max_line_length: Optional[List] = None,
+) -> None:
+    cmd = [
+        "checkpatch.pl",
+        "--strict",  # Be more annoying
+        "--no-tree",  # No kernel source tree present
+        # Some stuff to make parsing easier (see below):
+        "--emacs",
+        "--terse",
+        "--showfile",
+        "--no-summary",
+        "--color=never",
+        "--no-signoff",  # Don't require Signed-Off-By
+    ]
+
+    if errors_ignored is not None:
+        cmd += ["--ignore", ",".join(errors_ignored)]
+    elif errors_enabled is not None:
+        cmd += ["--types", ",".join(errors_enabled)]
+    if max_line_length is not None:
+        cmd += [f"--max-line-length={max_line_length}"]
+
+    result = run(cmd, input=commit_patch.encode(), stdout=PIPE)
```

Comment:
Would `check_output()` be appropriate here?

Reply:
No, we need to catch and read errors.